### PR TITLE
send-transaction-service metrics improvement

### PR DESCRIPTION
**Pull request by lijunwangs**
Use data_point report metrics  periodically.

### CODE REVIEW

File: send-transaction-service/src/send_transaction_service.rs [hunk 1]
```diff
@@ -257,6 +388,8 @@ impl SendTransactionService {
             .name("send-tx-receive".to_string())
             .spawn(move || loop {
                 let recv_timeout_ms = config.batch_send_rate_ms;
+                let mut stats = SendTransactionServiceStats::default();
+                let mut stats_updated = false;
```
Comment (t-nelson) [inline send-transaction-service/src/send_transaction_service.rs:392]: won't need to track this with atomics

File: send-transaction-service/src/send_transaction_service.rs [hunk 2]
```diff
@@ -519,54 +646,51 @@ impl SendTransactionService {
...
+            warn!(
+                "Failed to send transaction transaction to {}: {:?}",
+                tpu_address, err
+            );
+            stats.send_failure_count += 1;
         }
+
+        measure.stop();
+        stats.send_us += measure.as_us();
+        stats.send_count += 1;
```
Comment (t-nelson) [inline send-transaction-service/src/send_transaction_service.rs:699]: send _attempt_ count ?

File: send-transaction-service/src/send_transaction_service.rs [hunk 3]
```diff
@@ -181,6 +183,130 @@ where
...
+                ("retries", self.stats.retries, i64),
+                ("failed-tx", self.stats.failed_transactions, i64)
+            );
+        }
+    }
+}
+
+#[derive(Clone)]
+struct SendTransactionServiceStatsReporter {
+    report: Arc<Mutex<SendTransactionServiceStatsReport>>,
```
[the remark this answers is not in the document]
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs:274]: Good point. Working on it.

[the remark this answers is not in the document]
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs]: Done

File: send-transaction-service/src/send_transaction_service.rs (hunk 1 above)
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs]: Done

File: send-transaction-service/src/send_transaction_service.rs [hunk 4]
```diff
@@ -181,6 +183,130 @@ where
...
+                ("rooted-tx", self.stats.rooted_transactions, i64),
+                ("expired-tx", self.stats.expired_transactions, i64),
+                (
+                    "max-retries-exceeded-tx",
+                    self.stats.transactions_exceeding_max_retries,
+                    i64
+                ),
+                ("retries", self.stats.retries, i64),
+                ("failed-tx", self.stats.failed_transactions, i64)
+            );
```
[the remark this answers is not in the document]
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs:323]: Done

File: send-transaction-service/src/send_transaction_service.rs [hunk 5]
```diff
@@ -181,6 +184,150 @@ where
...
+
+    /// Count of transactions not being added to retry queue
+    /// due to queue size limit
+    retry_queue_overflow: AtomicU64,
+
+    /// retry queue size
+    retry_queue_size: AtomicU64,
+
+    /// The count of batches
+    send_count: AtomicU64,
```
Comment (CriesofCarrots) [inline send-transaction-service/src/send_transaction_service.rs:207]: It is not clear to me from the variable/metric name or the doc whether this is the number of batches sent, or number of transactions in batches.

File: send-transaction-service/src/send_transaction_service.rs [hunk 6]
```diff
@@ -181,6 +184,150 @@ where
...
+
+    /// Time spent on transactions in non-batch mode in micro seconds
+    send_us: AtomicU64,
```
Comment (CriesofCarrots) [inline send-transaction-service/src/send_transaction_service.rs:210]: Ha, sometimes the `us` abbreviation is really unfortunate. Nothing to do here; I just think `send_us` is funny

File: send-transaction-service/src/send_transaction_service.rs [hunk 7]
```diff
@@ -181,6 +184,150 @@ where
...
+                    self.stats.failed_transactions.swap(0, Ordering::Relaxed),
+                    i64
+                )
+            );
+        }
+    }
+}
+
+/// Report the send transaction memtrics for every 5 seconds.
+pub const SEND_TRANSACTION_METRICS_REPORT_RATE_MS: u64 = 5000;
```
Comment (CriesofCarrots) [inline send-transaction-service/src/send_transaction_service.rs:329]: Does this need to be pub?

File: send-transaction-service/src/send_transaction_service.rs [hunk 8]
```diff
@@ -288,20 +443,22 @@ impl SendTransactionService {
...
                         leader_info_provider.lock().unwrap().get_leader_info(),
                         &config,
+                        stats,
                     );
                     let last_sent_time = Instant::now();
                     {
                         // take a lock of retry_transactions and move the batch to the retry set.
                         let mut retry_transactions = retry_transactions.lock().unwrap();
+                        let txns_to_retry = transactions.len();
+                        let mut txns_added_to_retry = 0;
```
Comment (CriesofCarrots) [inline send-transaction-service/src/send_transaction_service.rs:461]:
```suggestion
                        let transactions_to_retry = transactions.len();
                        let mut transactions_added_to_retry = 0;
```

File: send-transaction-service/src/send_transaction_service.rs (hunk 5 above)
Comment (CriesofCarrots) [reply send-transaction-service/src/send_transaction_service.rs:207]: The doc should maybe also answer the question: does this include leader fanout?

Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs]: These are only send transaction calls -- not the count of transactions sent -- we already have sent_transactions. I updated the doc a little.

File: send-transaction-service/src/send_transaction_service.rs (hunk 2 above)
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs]: renamed to send_attemp_count

File: send-transaction-service/src/send_transaction_service.rs [hunk 9]
```diff
@@ -311,14 +468,22 @@ impl SendTransactionService {
                                     break;
                                 } else {
                                     transaction_info.last_sent_time = Some(last_sent_time);
+                                    txns_added_to_retry += 1;
                                     entry.or_insert(transaction_info);
                                 }
                             }
                         }
+                        stats.retry_queue_overflow.fetch_add(
+                            (txns_to_retry - txns_added_to_retry) as u64,
```
Comment (t-nelson) [inline send-transaction-service/src/send_transaction_service.rs:477]:
```suggestion
                            txns_to_retry.saturating_sub(txns_added_to_retry) as u64,
```

File: send-transaction-service/src/send_transaction_service.rs [hunk 10]
```diff
@@ -311,14 +468,22 @@ impl SendTransactionService {
                                     break;
                                 } else {
                                     transaction_info.last_sent_time = Some(last_sent_time);
+                                    txns_added_to_retry += 1;
```
Comment (t-nelson) [inline send-transaction-service/src/send_transaction_service.rs:471]:
```suggestion
                                    saturating_add_assign!(txns_added_to_retry, 1);
```

**Review by t-nelson — APPROVED** [review]
I'm down to nits. r+ Tyera's approval

File: send-transaction-service/src/send_transaction_service.rs (hunk 6 above)
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs:210]: Yes!

File: send-transaction-service/src/send_transaction_service.rs (hunk 7 above)
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs]: Removed pub. I was mulling over if this is public info. It is to understand there report rate. Then I thought a lot Internal implementation probably fall into that same reasoning... So good documentation is enough.

File: send-transaction-service/src/send_transaction_service.rs (hunk 8 above)
Comment (lijunwangs) [reply send-transaction-service/src/send_transaction_service.rs]: Done

**Pull request merged**